From ce841620f8c66a418fe9f2002960cfd6207cade4 Mon Sep 17 00:00:00 2001
From: Sudeep Biswas <sudeep.biswas@st.com>
Date: Thu, 28 Apr 2016 14:26:38 +0530
Subject: [PATCH v1.0[nfc-linux] 2/2] drivers: nfc: nfcst: Proper Type2
 ACK-NAK handling

Modification to handle ACK/NACK response from type2
tag in CR95HF driver and to leave CRC handling for
digital core framework in case of Type2 tags rather
than handling it in CR95HF driver.

Signed-off-by: Shikha Singh <shikha.singh@st.com>
Signed-off-by: Sudeep Biswas <sudeep.biswas@st.com>
---
 drivers/nfc/nfcst/core.c |   14 ++++++++++++++
 1 files changed, 14 insertions(+), 0 deletions(-)

diff --git a/drivers/nfc/nfcst/core.c b/drivers/nfc/nfcst/core.c
index 3bdd8b9..d05bf52 100644
--- a/drivers/nfc/nfcst/core.c
+++ b/drivers/nfc/nfcst/core.c
@@ -69,6 +69,7 @@
 #define DEVICE_TIMEOUT_ERROR		0x87
 #define DEVICE_NFCA_CRC_ERR_MASK	0x20
 #define DEVICE_NFCB_CRC_ERR_MASK	0x01
+#define TYPE2_RESP_ACK_NACK		0x90

 /* device transmission flag values */
 #define TRFLAG_NFCA_SHORT_FRAME		0x07
@@ -597,6 +598,11 @@ static int nfcst_error_handling(struct nfcst_context *context,
 	/* Check for CRC err only if CRC is present in the tag response */
 	switch (context->current_rf_tech) {
 	case NFC_DIGITAL_RF_TECH_106A:
+		/* In case of Type2 ACK and NACK response, no CRC check*/
+		if (context->ddev->curr_protocol == NFC_PROTO_MIFARE){
+			if(skb_resp->data[0] == TYPE2_RESP_ACK_NACK)
+				break;
+		}
 		if (context->sendrcv_trflag == TRFLAG_NFCA_STD_FRAME_CRC) {
 			error_byte = skb_resp->data[res_len - 3];
 			if (error_byte & DEVICE_NFCA_CRC_ERR_MASK) {
@@ -649,6 +655,14 @@ static void nfcst_response_handler(struct nfcst_context *context,
 	}
 	cb_arg->rats = false;

+	/** CRC handling **/
+	/* For MIFARE Type2, CRC handling will be done by digital framework */
+	if (context->current_rf_tech == NFC_DIGITAL_RF_TECH_106A &&
+	    context->ddev->curr_protocol == NFC_PROTO_MIFARE) {
+		/* Removing Transceiver specific data */
+		skb_trim(skb_resp, (skb_len - 3));
+		return;
+	}
 	/* Remove CRC bytes only if received frames data has an eod (CRC) */
 	switch (context->current_rf_tech) {
 	case NFC_DIGITAL_RF_TECH_106A:
--
1.7.5.4
